substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/attach/AttachListenerThread.java

```diff
@@ -132,20 +148,245 @@ private static void handleException(AttachOperation op, Throwable e) {
+        Pointer buffer = NativeMemory.malloc(bufferSize, NmtCategory.Serviceability);
+        try {
+            return readRequestData0(channel, buffer, bufferSize, minStringCount, minReadSize);
+        } finally {
+            NativeMemory.free(buffer);
+        }
+    }
+
+    private static AttachOperation readRequestData0(AttachSocketChannel channel, Pointer buffer, int bufferSize, int minStringCount, int minReadSize) {
+        int strCount = 0;
+        int offset = 0;
+        int left = bufferSize;
+
+        /*
+         * Read until all (expected) strings or expected bytes have been read, the buffer is full,
+         * or EOF.
+         */
+        do {
+            int n = channel.read(buffer.add(offset), left);
+            if (n < 0) {
+                return null;
+            } else if (n == 0) {
+                break;
+            }
+
+            if (minStringCount > 0) {
+                /* Need to count arguments. */
+                for (int i = 0; i < n; i++) {
+                    if (buffer.readByte(offset + i) == 0) {
+                        strCount++;
+                    }
+                }
+            }
+            offset += n;
+            left -= n;
+        } while (left > 0 && (offset < minReadSize || strCount < minStringCount));
+
+        if (offset < minReadSize || strCount < minStringCount) {
+            /* Unexpected EOF. */
+            return null;
+        } else if (buffer.readByte(offset - 1) != 0) {
+            /* Request must end with '\0'. */
+            return null;
+        }
+
+        /* Parse all strings. This part is very different from HotSpot. */
+        ArrayList<String> values = decodeStrings(buffer, buffer.add(offset));
+        return createAttachOperation(channel, values);
+    }
+
+    private static ArrayList<String> decodeStrings(Pointer dataStart, Pointer dataEnd) {
+        ArrayList<String> result = new ArrayList<>(4);
+
+        Pointer currentStart = dataStart;
+        Pointer pos = dataStart;
+        while (pos.belowThan(dataEnd)) {
+            if (pos.readByte(0) == 0) {
+                String s = decodeString(currentStart, pos);
+                result.add(s);
+
+                currentStart = pos.add(1);
+            }
+
+            pos = pos.add(1);
+        }
+
+        return result;
+    }
+
+    private static String decodeString(Pointer start, Pointer end) {
+        assert end.aboveOrEqual(start);
+        UnsignedWord length = end.subtract(start);
+        return CTypeConversion.toJavaString((CCharPointer) start, length, StandardCharsets.UTF_8);
+    }
+
+    private static AttachOperation createAttachOperation(AttachSocketChannel channel, ArrayList<String> values) {
+        /* Parse the name and the options. */
+        String name;
+        String nameAndOptions = values.getFirst();
+
+        int optionStart = nameAndOptions.indexOf(' ');
+        if (optionStart != -1) {
+            name = nameAndOptions.substring(0, optionStart);
```

Comment:
It doesn't hurt to do this check, but I think we can just assume there will never be any options specified by the client. Because at this point it has [already asked the server "getversion"](https://github.com/alexmenkov/jdk/blob/master/src/jdk.attach/linux/classes/sun/tools/attach/VirtualMachineImpl.java#L127) and knows we don't support options, so [it won't supply any](https://github.com/openjdk/jdk/pull/23405/files#diff-f90f733fa4c70b9b92e40856d5ec7a287f9586fbde49148ce986316e29aceabcR544).

Reply:
Good point - I will probably keep the check for now though (it makes the diff against HotSpot a bit smaller).